core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/UpgradeService.java

```diff
@@ -129,6 +140,47 @@ private UpgradeOptionV4Response getUpgradeOption(Stack stack, Long workspaceId,
         }
     }
 
+    private Predicate<com.sequenceiq.cloudbreak.cloud.model.catalog.Image> getImageFilter(Image image, boolean baseImage, Stack stack) {
+        if (baseImage) {
+            return packageVersionFilter(image.getPackageVersions());
+        } else {
+            return packageVersionFilter(image.getPackageVersions())
+                    .and(parcelFilter(stack));
+        }
+    }
+
+    private Predicate<com.sequenceiq.cloudbreak.cloud.model.catalog.Image> parcelFilter(Stack stack) {
+        Set<String> parcels = clusterComponentConfigProvider.getClouderaManagerProductDetails(stack.getCluster().getId())
+                .stream()
+                .map(ClouderaManagerProduct::getParcel)
+                .collect(toSet());
+        return imageFromCatalaog -> {
+            if (imageFromCatalaog.getPreWarmParcels() == null) {
+                return false;
+            } else {
+                Set<String> imageParcels = imageFromCatalaog.getPreWarmParcels()
```

Comment:
you can use `CollectionUtils.emptyIfNull()` here and then you do not need the `if-else` statement above